packages/lesswrong/lib/collections/posts/annualReviewMarkets.ts

```diff
@@ -152,15 +152,30 @@ export const createManifoldMarket = async (question: string, descriptionMarkdown
 
 
 
-async function refreshMarketInfoInCache(post: DbPost, context: ResolverContext) {
-  const marketInfo = await postGetMarketInfoFromManifold(post);
-  if (!marketInfo || !post.manifoldReviewMarketId) return null;
+//export async function testRefreshMarketInfoInCache(marketId: string, year: number) {
+//  const context = createAnonymousContext();
+//  await refreshMarketInfoInCache(marketId, year, context);
+//}
+
+async function refreshMarketInfoInCache(marketId: string, year: number, context: ResolverContext) {
+  // Update the market-info cache for a Manifold prediction market. In order to
+  // avoid thundering-herd issues, we update the cache item timestamp first (and
+  // check that it changed by a minimum amount) before we send the API request
+  // to Manifold.
+  const previousTimestamp = await context.repos.manifoldProbabilitiesCachesRepo.updateMarketInfoCacheTimestamp(marketId);
+  if (previousTimestamp && (new Date().getTime() - previousTimestamp.getTime() < 5_000)) {
```

Comment:
Doesn't this have the problem that if a post gets a sequence of requests where each subsequent request is not more than 5 seconds after the previous request, we could have an extended period of a cache record not being updated?  (I guess maybe individual posts don't generally see enough traffic for that to be an issue?)

Separately looks like we're doubling the number of writes to our own database in the cases where that doesn't happen (which should be most of them).  Though we're also extending the TTL by more than 2x, so it's probably fine?

Reply:
Oh, I see, i.e. if we get the same post multiple times either within the same request, or as a result of a single request doing some kind of fanout. 

But each one still performs an update operation, right? Though it's likely that the update is noticeably cheaper on the postgres side than the INSERT with the ON CONFLICT, due to locking semantics... so I guess it's a win in situations with higher volume, which might be all of them if we reliably get multiple concurrent upserts on a single post whenever its cache entry expires.